FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,61 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 	
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				if let cache = try CoreDataFeed.find(in: context) {
+					completion(.found(
+						feed: cache.items
+							.compactMap { ($0 as? CoreDataFeedImage)?.toLocal() },
+						timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 	
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				let managedCache = try CoreDataFeed.newInstance(in: context)
+				managedCache.timestamp = timestamp
+				managedCache.items = NSOrderedSet(array: feed.map { local in
+					let managed = CoreDataFeedImage(context: context)
+					managed.id = local.id
+					managed.descriptionText = local.description
+					managed.location = local.location
+					managed.url = local.url
+					return managed
+				})
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 	
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				if let cache = try CoreDataFeed.find(in: context) {
+					context.delete(cache)
+				}
+				
+				try context.save()
```

Comment:
The `save` can be inside the if-statement because it should only need to save if the cache was deleted.

```suggestion
				if let cache = try CoreDataFeed.find(in: context) {
					context.delete(cache)
					try context.save()
				}				
```

Reply:
If I do so the test `test_delete_deliversErrorOnDeletionError` fails because the `context.save()` is not called anymore and no error is thrown anymore in the test.